parser-typechecker/src/Unison/Codebase/Branch.hs

```diff
@@ -455,6 +459,35 @@ data ForkFailure = SrcNotFound | DestExists
 numHashChars :: Branch m -> Int
 numHashChars _b = 3
 
+-- This type is a little ugly, so we wrap it up with a nice type alias for
+-- use outside this module.
+type Cache m = Cache.Cache m (Causal.RawHash Raw) (Causal m Raw (Branch0 m))
+
+boundedCache :: MonadIO m => Word -> m (Cache m)
+boundedCache = Cache.semispaceCache
+
+-- todo: can delete old `read` once we are confident in this
+cachedRead :: forall m . Monad m
+           => Cache m
+           -> Causal.Deserialize m Raw Raw
+           -> (EditHash -> m Patch)
+           -> Hash
+           -> m (Branch m)
+cachedRead cache deserializeRaw deserializeEdits h =
+ Branch <$> Causal.cachedRead cache d h
+ where
+  fromRaw :: Raw -> m (Branch0 m)
+  fromRaw Raw {..} = do
+    children <- traverse go _childrenR
+    edits <- for _editsR $ \hash -> (hash,) . pure <$> deserializeEdits hash
+    pure $ branch0 _termsR _typesR children edits
+  go = cachedRead cache deserializeRaw deserializeEdits
+  d :: Causal.Deserialize m Raw (Branch0 m)
+  d h = deserializeRaw h >>= \case
+    RawOne raw      -> RawOne <$> fromRaw raw
+    RawCons  raw h  -> flip RawCons h <$> fromRaw raw
+    RawMerge raw hs -> flip RawMerge hs <$> fromRaw raw
```

Comment:
This ended up almost identical to existing `Branch.read`, just with `Causal.cachedRead` in place of `Causal.read`.

Reply:
Do we still use `Branch.read` anywhere?  (Not that it's necessarily time to delete it yet.)